usepackage.1.in
.TH USE 1 "$Date$" "Usepackage" "User Commands"
.SH NAME
usepackage \-
.I Usepackage
Environment Manager
.SH SYNOPSIS
See
.IR use(1) .
.SH DESCRIPTION
.I Usepackage
is an environment management program. It is based on the principle of
.I packages
\- collections of executables that share a common set of necessary
environment variables, such as PATH, MANPATH or LD_LIBRARY_PATH.
.PP
For each given
.IR package ,
.B use
sources the appropriate environment information into the current shell.
The environment information is specified in packages files, see
.IR PACKAGES .
.PP
The
.B usepackage
executable is the backend program used to generate environment information
suitable for sourcing into a running shell. The
.B use
frontend should be invoked to actually effect changes to the environment,
see
.IR use(1) .
.SS OPTIONS
.TP
.I \-v
Output verbose information to the standard error stream.
.TP
.I \-s
Silence warnings for un-matched packages. This is useful in a shell
.I rc
script when a package is known not to be available on all architectures that
the shell is used on, see
.IR PACKAGES .
.TP
.I \-c
Force
.I csh
style environment output.
.TP
.I \-b
Force bourne shell style environment output.
.TP
.I \-f file
Specify an alternate initial configuration file, see
.IR PACKAGES .
.TP
.I \-l
List available packages and groups, see
.IR ANNOTATIONS
and
.IR GROUPS .
.SH PACKAGES
.I Usepackage
reads package environment information from the configuration file as follows:
.SS LOAD ORDER
When
.I Usepackage
starts up, it loads an initial configuration file, the name of this file may
be given on the command line (see
.IR OPTIONS ),
otherwise it defaults to the builtin name "usepackage.conf" (see
.IR FILES ).
This file is located by searching along a path which defaults to:
.PP
.nf
     @sysconfdir@:~:.
.fi
.PP
This path may be overridden with the PACKAGES_PATH environment variable
(see
.IR ENVIRONMENT ).
A packages file may contain inclusion directives which cause the sourcing of
other files at given points. An inclusion directive looks like:
.PP
.nf
     (include file-name)
.fi
.PP
The same location mechanism is used to find "file-name" as for the initial
configuration file. A package file name may also be given as an absolute file
name or may be shell-style user directory, tilde (~), relative.
.SS SYNTAX
A package file consists of comments, delimited by a leading hash (#),
or package definitions of the form:
.PP
.nf
     package [arch [os [version [host [shell]]]]]
          [<= requires ...] : setting [, setting ...] ;
.fi
.PP
The
.IR package ,
.IR arch ,
.IR os ,
.IR version ,
.IR host ,
and
.I shell
parts may be simple shell-style patterns of the form:
.TP 15
*
matches anything.
.TP
foo*
matches "foo", "foobar" etc.
.TP
{foo,bar}
matches "foo" or "bar".
.PP
The
.I requires
list specifies the names of other packages which must be sourced into the
environment before the settings for this package are processed.
.PP
A
.I setting
is either a variable definition, an alias definition (both described below)
or a section of text to be directly evaluated in the shell, delimited with
.IR <[
and
.IR ]> ,
for example:
.PP
.nf
     <[ /usr/local/bin/test-init ]>
.fi
.PP
When
.I Usepackage
searches for the definition for a particular package, it compares each line
in the packages file against the name of the package given (package) and
system-dependant information for the execution host, as obtained by
.IR uname (2).
This information is the hardware implementation (platform), the operating
system name (os), the operating system version (version) and the hostname
(host). Comparisons are case-insensitive. If a match is obtained then the
given variable definitions and script sections are processed to modify the
environment. A  variable definition may have one of the following forms:
.PP
.nf
     var-name = "string"
     var-name = path-list
     var-name += path-list
     var-name += "string"
     var-name =+ path-list
     var-name =+ "string"
.fi
.PP
The first sets the given variable to a literal string value, the
second sets the given variable to a path list, the third prepends the
current value of a variable with the given path list, and the fourth
interprets the literal string as a path list and prepends it to the current
value of the variable. The fourth and fifth forms do the same, but append to
the current value of the variable. Path lists are colon (:) separated lists of
directories and may contain shell-style tilde (~), user-relative, directories
which will be expanded automatically (except when the path list is given as a
literal string). When pre-pending paths to a variable, duplicate paths are
removed from the original value first.
.PP
In addition,
.I Usepackage
can optionally test for the existence of paths before setting or adding them
to variables. This is done with the test form of the above operators:
.PP
.nf
     var-name ?= path-list
     var-name ?+= path-list
     var-name ?+= "string"
.fi
.PP
These operate as per the definitions above, but will ignore path components
that don't exist.
.PP
An alias definition has one of the following forms:
.PP
.nf
     alias name = name
     alias name = path
     alias name = "string"
.fi
.PP
These will define a matching shell alias.
.SS GROUPS
In addition to the package definitions in a packages file, there may also be
group definitions. These have the following syntax:
.PP
.nf
     group := package [, package ...] ;
.fi
.PP
.I Usepackage
searches for a given package name in the defined groups first, if the given
name matchs a group name then the packages defined as part of that group are
sourced into the environment together. A group definition may not reference
other groups and may not contain patterns.
.SS ANNOTATIONS
In order to give useful package information to the user, annotations may be
placed in the packages file that give summaries of packages. These annotations
have the form:
.PP
.nf
     >> name : "description" <<
.fi
.PP
.I Usepackage
collects these annotations together and displays them when called with the
.I -l
flag, see
.IR OPTIONS .
These annotations have no impact on the package mechanism and need not
necessarily be beside or correspond to the package definitions (although
this is the sensible way to arrange things).
.SS EXAMPLE
The following fragment of a packages file illustrates the main features:
.PP
.nf
     # GNU software is available everywhere:

     >> GNU : "The GNU project software" <<

     GNU :                     PATH += /usr/local/gnu/bin,
                               MANPATH += /usr/local/gnu/man ;

     # CVS requires RCS which is found in the GNU package, but
     # is only available on SPARC Solaris machines:

     >> CVS : "Concurrent Versions System revision control" <<

     CVS sun4* SunOS 5.* <= GNU :
                               CVSROOT = /usr/src/cvsroot,
                               CVSEDITOR = "vi",
                               PATH += /usr/local/cvs/bin,
                               MANPATH += /usr/local/cvs/man ;

     # User bin directories (Solaris will run SunOS 4 binaries):
     # (Only added to the path if the directories actually exist)

     >> user : "User's own programs" <<

     user sun4* SunOS :        PATH ?+= ~/bin/sun4 ;
     user sun4* SunOS 5.* :    PATH ?+= ~/bin/solaris ;
     user alpha OSF :          PATH ?+= ~/bin/alpha ;

     # Useful aliases:

     aliases : alias more = less,
               alias ls = "ls --color",
               alias cc = /usr/bin/gcc ;

     # Special function for zsh:

     >> zsh-function : "Special zsh function" <<

     zsh-function * * * * zsh :
           <[ hello() { echo "Hello World!"; } ]> ;

     # include standard packages:

     (include packages.standard)

     # security hole:

     dot :                     PATH += . ;

     # groups:

     user-setup := standard, user ;
     programmer-setup := standard, CVS, user, dot ;
.fi
.PP
Note the use of Operating System version numbers to distinguish between
SunOS 4 and Solaris (SunOS 5), the use of "sun4*" to match the multiple
different platform versions of SPARC machines (sun4m, sun4c, etc.) and the
fact that package "user" on a SPARC Solaris machine will match both of the
first two lines of the "user" package section, resulting in the "solaris"
directory and the "sun4" directory being added into the PATH.
.PP
Assuming the appropriate shell setup script has been sourced - see
.IR use(1)
- then the following command will cause the CVS environment (including the
GNU environment) to be sourced:
.PP
.nf
     $ use CVS
.fi
.PP
Note that on a DEC Alpha machine, this will generate a warning like:
.PP
.nf
     $ use CVS
     warning: no match for package `CVS' on this host.
.fi
.PP
In a shell script which is executed on a number of different platforms (such
as the shell startup script). These warnings may be silenced (see
.IR OPTIONS ).
.SH FILES
.TP 15
.SB @pkgdatadir@/usepackage.conf
The default main packages file.
.TP
.SB @pkgdatadir@/use.csh
Shell setup for csh and derivatives.
.TP
.SB @pkgdatadir@/use.bsh
Shell setup for bourne shell and derivatives.
.TP
.SB @pkgdatadir@/use.ksh
Shell setup for ksh.
.TP
.SB @bindir@/usepackage
The underlying
.I Usepackage
executable.
.SH ENVIRONMENT
Other than the reading and re-definition of environment variables for package
setup,
.B use
also uses the following environment variables for user configuration:
.TP 15
.SB PACKAGES_PATH
Colon-separated path list giving the directories to search for configuration
files. Shell-style tilde (~) user-directory escapes are expanded.
.TP
.SB HOME
If present in the environment, this is used to provide the expansion for
the tilde (~) user-directory.
.TP
.SB SHELL
If present in the environment, the last path component of this is used for
shell matching (see
.IR SYNTAX )
and detecting the style of environment output that should be used (see
.IR OPTIONS ).
.SH COPYRIGHT
.nf
Usepackage Environment Manager
Copyright (C) 1995-2020  Jonathan Hogg  <me@jonathanhogg.com>
.fi
.PP
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.
.PP
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
.PP
You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
.SH SEE ALSO
use(1), csh(1), sh(1), ksh(1), env(1), environ(5), getenv(3C), uname(1),
uname(2)


NEWS
Usepackage Environment Manager
Copyright (C) 1995-2020  Jonathan Hogg  <me@jonathanhogg.com>

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

Changes between usepackage version 1.13 and 1.14

* Just updating copyrights on everything again for the 25TH ANNIVERSARY!!!
  You may be amazed to hear that I still use this tool on a daily basis.

Changes between usepackage version 1.11 and 1.13

* Just updating the email addresses and copyrights on everything for the
  20TH ANNIVERSARY!!!

Changes between usepackage version 1.11 and 1.12

* More code updating, this time to fix an inclusion bug related to very
  old-school usage of flex.

Changes between usepackage version 1.10 and 1.11

* The code was updated a fair bit to match more modern autoconf/automake
  and C99 standards (it's really starting to show its age).

Changes between usepackage version 1.8 and 1.10

* Now supports shell-independent aliases using 'alias foo = bar' syntax.
* Now supports adding path segments at the end using the new "=+"
  syntax (courtesy of Chandler Wilkerson).
* Skipping 1.9 version number as I never did a real release of it and
  I added the alias stuff afterwards anyway.

Changes between usepackage version 1.7 and 1.8

* Now supports basic checking for existence of directories before adding
  them to a path (or setting a variable) using the new "?=" and "?+="
  operators.
* The order that variables are sourced into the environment has been
  changed to better support using other variables as parts of updates
  to the environment (e.g., "FOO=/opt/foo, PATH+=$FOO/bin;").

Changes between usepackage version 1.6.3 and 1.7

* Now supports arbitrary inclusions into the shell evaluation text,
  allowing commands to be executed, and shell aliases and functions
  to be defined.
* Match patterns (other than the initial package name) are now optional,
  such that a missing pattern defaults to a "*" wildcard.
* An additional optional pattern has been added to match the user's
  shell. This is determined from $SHELL or the shell defined in the
  password database for the running user.

Changes between usepackage version 1.6.2 and 1.6.3

* Fixed ~ expansion to use $HOME if available.

Changes between usepackage version 1.6.1 and 1.6.2

* Split manual page into two (for frontend and backend).
* Cosmetic changes.

Changes between usepackage version 1.6 and 1.6.1

* Now uses autoconf/automake for the build system.
* Small changes to the output messages.
* Fixed up the contact information.
* A bit more portable hopefully.

Initial public release 1.6

* Usepackage has been in use for some time, and had even been available
  for download, but no attempt was made previously to publicise it. This
  release represents the first posted to freshmeat.net.


usepackage.c

/*****************************************************************************
 *
 * Usepackage Environment Manager
 * Copyright (C) 1995-2020  Jonathan Hogg  <me@jonathanhogg.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 * Name   : usepackage.c
 * Author : Jonathan Hogg <me@jonathanhogg.com>
 *
 ****************************************************************************/


/* usepackage.c */

/* Jonathan AH Hogg */


/*** uses: ***/

#include "config.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/utsname.h>
#include <dirent.h>
#include "linked_list.h"
#include "package.h"
#include "utils.h"


/*** prototypes: ***/

void add_package(package_t* package);
void use_package(char* name);
void use_group(group_t* group);
group_t* get_group(char* name);
void print_env(void);
list_node* get_into_env(variable_t* var);
linked_list* make_pathlist(char* path_string);
variable_t* update_var(variable_t* evar, variable_t* vvar);
void print_path(char* varname, linked_list* pathlist);
linked_list* merge_paths(linked_list* elist, linked_list* vlist);
linked_list* merge_paths_appended(linked_list* elist, linked_list* vlist);
linked_list* test_paths(linked_list* vlist);
void list_annotations();
void list_groups();
void print_scripts(void);
void print_aliases(void);


/*** globals: ***/

int debugging = 0;
int silent = 0;
int csh_user;
struct utsname the_host_info;
char* the_shell;
linked_list* the_packages;
linked_list* the_groups;
linked_list* the_annotations;
linked_list* the_environment;
linked_list* the_scripts;
linked_list* the_aliases;
char* main_package_filename = MAIN_PACKAGE_FILE;


/*** main program: ***/

int main(int argc, char *argv[])
{
   int first,i;
   char* f;
   group_t* group;
   int sh_override = -1;
   int list_packages = 0;

   for (i=1; i<argc && *argv[i] == '-'; i++)
   {
      for (f=argv[i]+1; *f; f++)
         switch (*f)
         {
            case 'v':
               debugging = 1;
               silent = 0;
               break;
            case 's':
               debugging = 0;
               silent = 1;
               break;
            case 'c':
               sh_override = 1;
               break;
            case 'b':
               sh_override = 0;
               break;
            case 'f':
               main_package_filename = argv[++i];
               break;
            case 'l':
               list_packages = 1;
               break;
            default:
               fprintf(stderr, "usepackage: unrecognised flag '%c'\n", *f);
               exit(1);
         }
   }

   if (!list_packages && (i >= argc))
   {
      fprintf(stderr, "\n%s %s, %s\n", PACKAGE_NAME, PACKAGE_VERSION, COPYRIGHT);
      fprintf(stderr, "Bug reports and comments to: %s\n", PACKAGE_BUGREPORT);
      fprintf(stderr, "usage: use [-vs] [-f <file>] <package> [<package>...]\n");
      fprintf(stderr, "       use -l\n\n");
      fprintf(stderr, "       -v : verbose\n");
      fprintf(stderr, "       -s : silence match warnings\n");
      fprintf(stderr, "       -f : use <file> as main packages file\n");
      fprintf(stderr, "       -l : list available packages\n\n");
      exit(1);
   }

   DEBUG(stderr, "%s\n", PACKAGE_NAME);
   DEBUG(stderr, "Version: %s\n", PACKAGE_VERSION);
   DEBUG(stderr, "%s\n", COPYRIGHT);

   uname(&the_host_info);
   DEBUG(stderr, "host: %s\n", the_host_info.nodename);
   DEBUG(stderr, "operating system: %s %s\n", the_host_info.sysname,
         the_host_info.release);
   DEBUG(stderr, "architecture: %s\n", the_host_info.machine);

   the_shell = get_user_shell();
   DEBUG(stderr, "shell: %s\n", the_shell);

   if (sh_override != -1)
      csh_user = sh_override;
   else
      csh_user = ((!strcmp(the_shell, "csh")) || (!strcmp(the_shell, "tcsh")));

   the_environment = new_list();
   the_scripts = new_list();
   the_aliases = new_list();

   if (get_packages(&the_packages, &the_groups, &the_annotations))
   {
      fprintf(stderr, "usepackage: couldn't load package information.\n");
      exit(2);
   }

   if (list_packages)
   {
      fprintf(stderr, "\nusepackage %s, %s\n\n", VERSION, COPYRIGHT);
      fprintf(stderr, "Available packages are:\n\n");
      list_annotations();
      fprintf(stderr, "\nAvailable groups are:\n\n");
      list_groups();
      fprintf(stderr, "\n");
      exit(0);
   }

   for (first = i, i = argc - 1 ; i >= first ; i--)
   {
      if ((group = get_group(argv[i])))
         use_group(group);
      else
         use_package(argv[i]);
   }

   print_env();

   print_scripts();

   print_aliases();

   return(0);
}


/*** functions: ***/

void use_package(char* name)
{
   package_t* package;
   list_node* node;
   int got_one = 0;

   DEBUG(stderr, "using package `%s'...\n", name);

   for (node = head(the_packages) ; node ; node = next(node))
   {
      package = (package_t*) get_value(node);

      if (package_matches(package, name, the_host_info.machine,
                          the_host_info.sysname, the_host_info.release,
                          the_host_info.nodename, the_shell))
      {
         add_package(package);
         got_one = 1;
      }
   }

   if ((!silent) && (!got_one))
      fprintf(stderr,
              "warning: no match for package `%s' on this host.\n",
              name);
}

void add_package(package_t* package)
{
   variable_t* evar;
   variable_t* vvar;
   list_node* enode;
   list_node* vnode;
   list_node* rnode;
   group_t* group;
   char* name;
   int got_one;
   char* text;

   if (package->requires)
   {
      DEBUG(stderr, "(pre-using required packages list)\n");

      for (rnode=list_tail(package->requires) ; rnode ; rnode=previous(rnode))
      {
         name = (char*) get_value(rnode);

         if ((group = get_group(name)))
            use_group(group);
         else
            use_package(name);
      }
   }

   for (vnode = head(package->variables) ; vnode ; vnode = next(vnode))
   {
      vvar = get_value(vnode);
      got_one = 0;
      for (enode = head(the_environment) ; enode ; enode = next(enode))
      {
         evar = get_value(enode);
         if (!strcmp(vvar->name, evar->name))
         {
            remove_node(the_environment, enode, 0);
            add_to_tail(the_environment, update_var(evar, vvar));
            got_one = 1;
            break;
         }
      }
      if (!got_one)
      {
         enode = get_into_env(vvar);
         evar = get_value(enode);
         set_value(enode, update_var(evar, vvar));
      }
   }

   for (vnode = head(package->scripts) ; vnode ; vnode = next(vnode))
   {
      text = ((script_t*) get_value(vnode))->text;
      add_to_tail(the_scripts, (void*) text);
   }

   for (vnode = head(package->aliases) ; vnode ; vnode = next(vnode))
   {
      add_to_tail(the_aliases, get_value(vnode));
   }
}

void use_group(group_t* group)
{
   list_node* node;

   DEBUG(stderr, "(expanding group `%s')\n", group->name);

   for (node = list_tail(group->packages) ; node ; node = previous(node))
      use_package((char*) get_value(node));
}

group_t* get_group(char* name)
{
   list_node* node;
   group_t* group;

   for (node = head(the_groups) ; node ; node = next(node))
   {
      group = (group_t*) get_value(node);
      if (!strcasecmp(name, group->name))
         return(group);
   }

   return(NULL);
}

void print_env(void)
{
   list_node* node;
   variable_t* var;

   DEBUG(stderr, "dumping environment...\n");

   for (node = head(the_environment) ; node ; node = next(node))
   {
      var = (variable_t*) get_value(node);
      switch (var->type)
      {
         case VAR_LIT_SET:
            if (var->literal)
            {
	       if (csh_user)
		  printf("setenv %s \"%s\";\n", var->name, var->literal);
	       else
		  printf("%s=\"%s\";\nexport %s;\n", var->name, var->literal, var->name);
            }
            break;

         case VAR_PATH_ADD:
         case VAR_PATH_SET:
            if (var->pathlist)
            {
	       if (csh_user)
	       {
		  printf("setenv %s ", var->name);
		  print_path(var->name, var->pathlist);
		  printf(";\n");
	       }
	       else
	       {
		  printf("%s=", var->name);
		  print_path(var->name, var->pathlist);
		  printf(";\nexport %s;\n", var->name);
	       }
            }
            break;

         default:
            break;
      }
   }
}

void print_scripts(void)
{
   list_node* node;
   char* text;

   DEBUG(stderr, "dumping scripts...\n");

   for (node = head(the_scripts) ; node ; node = next(node))
   {
      text = (char*) get_value(node);
      printf("%s ;\n", text);
   }
}

void print_aliases(void)
{
   list_node* node;
   alias_t* alias;

   DEBUG(stderr, "dumping aliases...\n");

   for (node = head(the_aliases) ; node ; node = next(node))
   {
      alias = (alias_t*) get_value(node);
      if (csh_user)
      {
         printf("alias %s %s ;\n", alias->name, alias->text);
      }
      else
      {
         printf("alias %s=\"%s\" ;\n", alias->name, alias->text);
      }
   }
}

void print_path(char* varname, linked_list* pathlist)
{
   list_node* node;
   char* dirname;
   DIR* dir;

   /* special case for the empty path: */
   node = head(pathlist);
   dirname = get_value(node);
   if ((strlen(dirname) == 0) && !next(node))
   {
      printf("\"\"");
      return;
   }

   for (node = head(pathlist) ; node ; node = next(node))
   {
      dirname = get_value(node);
      if(strlen(dirname))
      {
        printf(next(node) ? "%s:" : "%s", dirname);
      }
   }
}

list_node* get_into_env(variable_t* var)
{
   variable_t* env_var;

   env_var = new(variable_t);
   env_var->name = var->name;
   env_var->type = var->type;

   switch(var->type)
   {
      case VAR_LIT_SET:
         env_var->literal = NULL;
         break;

      case VAR_PATH_SET:
         env_var->pathlist = NULL;
         break;

      case VAR_PATH_APPEND:
      case VAR_PATH_ADD:
         env_var->pathlist = make_pathlist(getenv(var->name));
         break;

      default:
         break;
   }

   return(add_to_tail(the_environment, env_var));
}

linked_list* make_pathlist(char* path_string)
{
   linked_list* list;
   int start,i,end;
   char* path;

   list = new_list();

   if (!path_string)
      return(list);

   start = 0;
   i = 0;
   end = strlen(path_string);

   while (i <= end)
   {
      if ((path_string[i] == ':') || (path_string[i] == '\0'))
      {
         path = (char*) malloc((i - start) + 1);
         strncpy(path, path_string + start, i - start);
         path[i - start] = '\0';
         add_to_tail(list, path);
         start = i + 1;
      }

      i++;
   }

   return(list);
}

variable_t* update_var(variable_t* evar, variable_t* vvar)
{
   linked_list* testlist;


   switch (vvar->type)
   {
      case VAR_LIT_SET:
         evar->literal = vvar->literal;
         evar->type = VAR_LIT_SET;
         break;

      case VAR_PATH_SET:
         evar->pathlist = vvar->pathlist;
         evar->type = VAR_PATH_SET;
         break;

      case VAR_PATH_ADD:
         switch (evar->type)
         {
            case VAR_LIT_SET:
               evar->pathlist = merge_paths(make_pathlist(evar->literal),
                                            vvar->pathlist);
               break;

            case VAR_PATH_SET:
            case VAR_PATH_ADD:
               evar->pathlist = merge_paths(evar->pathlist, vvar->pathlist);
               break;

            default:
               break;
         }
         evar->type = VAR_PATH_ADD;
         break;

       case VAR_PATH_APPEND:
         switch (evar->type)
         {
            case VAR_LIT_SET:
               evar->pathlist = merge_paths_appended(make_pathlist(evar->literal),
                                            vvar->pathlist);
               break;

            case VAR_PATH_SET:
            case VAR_PATH_ADD:
            case VAR_PATH_APPEND:
               evar->pathlist = merge_paths_appended(evar->pathlist, vvar->pathlist);
               break;

            default:
               break;
         }
         evar->type = VAR_PATH_ADD;
         break;

      case VAR_PATH_TESTSET:
         testlist = test_paths(vvar->pathlist);
         if (head(testlist))
         {
            evar->pathlist = testlist;
         }
         else
         {
            evar->pathlist = NULL;
         }
         evar->type = VAR_PATH_SET;
         break;

      case VAR_PATH_TESTADD:
         switch (evar->type)
         {
            case VAR_LIT_SET:
               evar->pathlist = merge_paths(make_pathlist(evar->literal),
                                            test_paths(vvar->pathlist));
               break;

            case VAR_PATH_SET:
            case VAR_PATH_ADD:
               evar->pathlist = merge_paths(evar->pathlist, test_paths(vvar->pathlist));
               break;

            default:
               break;
         }
         evar->type = VAR_PATH_ADD;
         break;

      default:
         break;
   }

   return(evar);
}

linked_list* merge_paths(linked_list* elist, linked_list* vlist)
{
   list_node* enode;
   list_node* vnode;
   char* vpath;

   if (elist && vlist)
   {
      for (vnode = list_tail(vlist) ; vnode ; vnode = previous(vnode))
      {
	 vpath = (char*) get_value(vnode);

	 for (enode = head(elist) ; enode ; enode = next(enode))
	    if (!strcmp(vpath, (char*) get_value(enode)))
	    {
	       remove_node(elist, enode, 0);
	       break;
	    }

	 add_to_head(elist, get_value(vnode));
      }
   }
   else if (vlist && head(vlist))
   {
      return(vlist);
   }

   return(elist);
}

linked_list* merge_paths_appended(linked_list* elist, linked_list* vlist)
{
   list_node* enode;
   list_node* vnode;
   char* vpath;


   if (elist && vlist)
   {
      for (vnode = head(vlist) ; vnode ; vnode = next(vnode))
      {
	 vpath = (char*) get_value(vnode);

	 for (enode = head(elist) ; enode ; enode = next(enode)) {
	    if (!strcmp(vpath, (char*) get_value(enode)))
	    {
	       remove_node(elist, enode, 0);
	       break;
	    }
	 }

	 add_to_tail(elist, get_value(vnode));

      }
   }
   else if (vlist && head(vlist))
   {
      return(vlist);
   }

   return(elist);
}

linked_list* test_paths(linked_list* vlist)
{
   linked_list* elist = new_list();
   list_node* vnode;
   char* vpath;
   FILE* test;

   for (vnode = list_tail(vlist) ; vnode ; vnode = previous(vnode))
   {
      vpath = (char*) get_value(vnode);

      test = fopen(vpath, "r");
      if ( test )
      {
         add_to_head(elist, vpath);
         fclose(test);
      }
   }

   return(elist);
}

void list_annotations()
{
   list_node* node;
   annotation_t* annotation;

   for (node = head(the_annotations); node; node = next(node))
   {
      annotation = (annotation_t*) get_value(node);

      fprintf(stderr, "   %s - %s\n", annotation->name,
              annotation->description);
   }
}

void list_groups()
{
   list_node* gnode;
   list_node* pnode;
   group_t* group;

   for (gnode = head(the_groups) ; gnode ; gnode = next(gnode))
   {
      group = (group_t*) get_value(gnode);

      fprintf(stderr, "  %s - ", group->name);

      for (pnode = head(group->packages) ; pnode ; pnode = next(pnode))
      {
         fprintf(stderr, "%s", (char*) get_value(pnode));
         if (next(pnode))
            fprintf(stderr, ", ");
      }

      fprintf(stderr, "\n");
   }
}
